### PR TITLE
Devel20220802

### PR DESCRIPTION
**Pull request by fepitre**
From @marmarek's review and suggestions:

- `fetch` plugin:
  - [x] isolate files verification (hash, signature) from download process (prevent download/decompress process compromise to bypass integrity check)
  - [x] distfiles dir -> per component

- Extra validation of `.qubesbuilder`:
  - [x] forbid ".." in "build", "url" and any other paths
  - [x] explicitly refuse files conflicting with stage info (e.g. package.fetch.yml)

- General code style comment:
  - [x] all the run() functions can be de-indented one level if you do `if stage != "build": return` (or whatever other stage)

- `build_rpm` and `sign_rpm`:
  - [x] add buildinfo

- `publish` stage:
  - [x] allow to configure `MIN_AGE_DAYS` (this will need to be adapted for builder-github)

### CODE REVIEW

**Comment by fepitre** [comment]
PipelineRetry

**Comment by fepitre** [comment]
PipelineRetry

**Comment by fepitre** [comment]
@marmarek You may want to review all of this, there is a lot of changes.

File: qubesbuilder/component.py [hunk 1]
```diff
@@ -29,9 +29,27 @@
...
+        for k, v in data.items():
+            deep_check(k)
+            deep_check(v)
+    elif isinstance(data, list):
+        for l in data:
+            deep_check(l)
+    elif isinstance(data, str):
+        for p in FORBIDDEN_PATTERNS:
+            if p in data:
+                raise ValueError(f"Forbidden pattern '{p}' found in '{data}'.")
```
Comment (marmarek) [inline qubesbuilder/component.py:51]: I'd prefer this to handle all possible (expected) types, and fail if unexpected type is found. IIUC it requires adding:
```
elif isinstance(data, int):
    pass
else:
    raise ValueError(f"Unexpected data type {type(data)} found")
```

File: qubesbuilder/component.py [hunk 2]
```diff
@@ -29,9 +29,27 @@
 import yaml
 from packaging.version import Version, InvalidVersion
 
-from qubesbuilder.common import sanitize_line
+from qubesbuilder.common import sanitize_line, STAGES
 from qubesbuilder.exc import ComponentError
 
+FORBIDDEN_PATTERNS = [".."]
+for s in STAGES:
+    FORBIDDEN_PATTERNS += [f".{s}.yml", f".{s}.yaml"]
```
Comment (marmarek) [inline qubesbuilder/component.py:37]: This is probably necessary too, but I think it should check for filename conflicts with anything that is in `copy_out`. This includes names of subpackages and any other file list that is determined dynamically (extend `is_filename_valid()`? and verify if all the places that dynamically build a filename use it). In fact `source_{rpm,deb}` may require that all subpackages have proper extension (`.deb`, `.rpm` etc).

**Pull request merged**